src/Build/Construction/ProjectElement.cs

```diff
@@ -460,6 +520,87 @@ protected internal virtual ProjectElement Clone(ProjectRootElement factory)
         /// <param name="owner">The factory to use for creating the new instance.</param>
         protected abstract ProjectElement CreateNewInstance(ProjectRootElement owner);
 
+        internal static ProjectElement CreateNewInstance(ProjectElement xml, ProjectRootElement owner)
+        {
+            if (xml.Link != null)
+            {
+                return xml.Link.CreateNewInstance(owner);
+            }
+
+            return xml.CreateNewInstance(owner);
+        }
+
+        internal ElementLocation GetAttributeLocation(string attributeName)
+        {
+            return Link != null ? Link.GetAttributeLocation(attributeName) : XmlElement.GetAttributeLocation(attributeName);
+        }
+
+        internal string GetAttributeValue(string attributeName, bool nullIfNotExists = false)
+        {
+            return Link != null ? Link.GetAttributeValue(attributeName, nullIfNotExists) :
+                ProjectXmlUtilities.GetAttributeValue(XmlElement, attributeName, nullIfNotExists);
+        }
+
+        internal string GetAttributeValue(string attributeName, ref string cache)
+        {
+            if (cache != null) return cache;
+            var value = GetAttributeValue(attributeName, false);
+            if (Link == null)
+            {
+                cache = value;
+            }
+
+            return value;
+        }
+
+        internal virtual void ClearAttributeCache()
+        {
+            this._condition = null;
+        }
+
+        internal void SetOrRemoveAttributeForLink(string name, string value, bool clearAttributeCache, string reason, string param)
+        {
+            SetOrRemoveAttribute(name, value, reason, param);
+            if (clearAttributeCache)
+            {
+                this.ClearAttributeCache();
+            }
+        }
+
+        internal void SetOrRemoveAttribute(string name, string value, string reason = null, string param = null)
+        {
+            if (Link != null)
+            {
+                Link.SetOrRemoveAttribute(name, value, false, reason, param);
+            }
+            else
+            {
+                ProjectXmlUtilities.SetOrRemoveAttribute(XmlElement, name, value);
+                if (reason != null)
+                {
+                    MarkDirty(reason, param);
+                }
+            }
+        }
+
+        internal void SetOrRemoveAttribute(string name, string value, ref string cache, string reason = null, string param = null)
+        {
+            if (Link != null)
+            {
+                Link.SetOrRemoveAttribute(name, value, true, reason, param);
```

Comment:
yes - thats why we pass the reason and param really, so we can generate the same "outcome" in real collection. Basically this will call the exact same method with exact samre params on the "other side" basically real object will move via "else" case there.